app-creator/client/src/widgets/template-wizard/template-wizard.ts

```diff
@@ -0,0 +1,614 @@
+   */
+  @property({ type: String }) deviceFilter: DeviceType = DeviceType.all;
+
+  /**
+   * Array of templates.
+   */
+  @property({ type: Array }) templates: TemplateItem[] = [];
+
+  /**
+   * Loading state.
+   */
+  @property({ type: Boolean }) loading = false;
+
+  /**
+   * Reference to dialog element.
+   */
+  @query('paper-dialog') dialog!: PaperDialogElement;
+
+  /**
+   * Reference to dialog element.
+   */
+  @query('#template-parsing-error') parsingErrorToast!: PaperToastElement;
+
+  /**
+   * ID of the currently selected template.
+   */
+  @property({ type: String }) selectedTemplateID = '';
+
+  /**
+   * Configuration object that stores metadata about a template.
+   */
+  @property({ type: Object }) config: { [key: string]: string } = {
+    id: generateRandomId(),
+    name: '',
+    palette: Palette.light,
+  };
+
+  async firstUpdated() {
+    try {
+      this.loading = true;
+
+      const templates = await templatesManager.fetchTemplates(false);
+      if (templates != null) {
+        this.templates = templates;
+      }
+    } catch (e) {
+      const fetchErrorToast = this.shadowRoot?.getElementById(
+        'fetch-error-toast'
+      ) as PaperToastElement;
+
+      if (fetchErrorToast != null) {
+        fetchErrorToast.open();
+      }
+      this.templates = templatesManager.getTemplates();
+    } finally {
+      this.loading = false;
+    }
+    this.showTemplateSelectionModal();
+  }
+
+  showTemplateSelectionModal() {
+    this.dialog.open();
+  }
+
+  /**
+   * Used for filtering templates.
+   */
+  handleDeviceFilters(device: DeviceType) {
+    this.deviceFilter = device;
+  }
+
+  /**
+   * Generates an input header for each setting input.
+   */
+  getInputHeader(title: string) {
+    return html`
+      <div class="input-header">
+        <p class="input-label">${title}</p>
+      </div>
+    `;
+  }
+
+  /**
+   * Sets currently selected template id.
+   */
+  handleTemplateSelection(id: string) {
+    this.selectedTemplateID = id;
+  }
+
+  /**
+   * Returns array of template cards.
+   */
+  getTemplateCards(showTitle = false): Array<TemplatesTabItem> {
+    return this.templates.map(({ id, name, imageUrl, device }) => {
+      return {
+        id,
+        name,
+        device,
+        markup: html`
+          ${showTitle ? nothing : html`<h6 class="subtitle">${name}</h6>`}
+          <template-card
+            id="${id}"
+            title="${name}"
+            imageUrl="${imageUrl}"
+            ?selected=${this.selectedTemplateID === id}
+            ?showTitle=${showTitle}
+            .onSelection=${() => this.handleTemplateSelection(id)}
+          ></template-card>
+        `,
+      };
+    });
+  }
+
+  /**
+   * Sets the query property when an onsearch event is dispatched from the
+   * searchbar widget.
+   */
+  handleSearch({ detail: { query } }: onSearchEvent) {
+    this.query = query.trim();
+  }
+
+  /**
+   * Creates a text input element.
+   */
+  getTextInput(
```

Comment:
nit: `getTextInput`, `getSelectInput`, and `getInputHeader` imply that they already exist. Perhaps these should be `make*`?